x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -286,18 +288,42 @@ static Set<String> fieldNames(LogicalPlan parsed, Set<String> enrichPolicyMatchF
         // otherwise, in some edge cases, we will fail to ask for "*" (all fields) instead
         references.removeIf(a -> a instanceof MetadataAttribute || MetadataAttribute.isSupported(a.qualifiedName()));
         Set<String> fieldNames = references.names();
+        Set<String> subFieldNames = new LinkedHashSet<>();
 
         if (fieldNames.isEmpty() && enrichPolicyMatchFields.isEmpty()) {
             // there cannot be an empty list of fields, we'll ask the simplest and lightest one instead: _index
             return IndexResolver.INDEX_METADATA_FIELD;
         } else {
-            fieldNames.addAll(subfields(fieldNames));
+            subFieldNames.addAll(subfields(fieldNames));
             fieldNames.addAll(enrichPolicyMatchFields);
-            fieldNames.addAll(subfields(enrichPolicyMatchFields));
+            subFieldNames.addAll(subfields(enrichPolicyMatchFields));
+            if (fieldNames.size() + subFieldNames.size() >= Operations.DEFAULT_DETERMINIZE_WORK_LIMIT * 2) {
+                Set<String> subfieldNamePrefixes = subfieldNamePrefixes(subFieldNames, 2);
+                if (fieldNames.size() + subfieldNamePrefixes.size() >= Operations.DEFAULT_DETERMINIZE_WORK_LIMIT * 2) {
+                    return IndexResolver.ALL_FIELDS;
```

Comment:
> Not a bad idea. Rather than rejecting the query because of the automaton limit, I _think_ we are now spending (a matter of seconds I think there were before we had the **new** IndexResolver that @nik9000 created) more time to get a rather large field_caps response from nodes and have a slower query.
> 
> I would also like to see a test that reveals this automaton limitation issue.

Thank you for reviewing @astefan! After discussing with @nik9000, I have #111118 to change `Regex.simpleMatchToAutomaton` to consolidate patterns end with ".*" into one Automaton, it affects more than ES|QL, but perhaps that is a better solution for this particular case comparing to what I proposed in this PR. I'm going to close this PR.